--- ./libs/libcurl/Makefile.orig	2016-06-20 16:35:50.719810919 +0300
+++ ./libs/libcurl/Makefile	2016-06-20 16:37:20.724068970 +0300
@@ -1,2 +1,3 @@
+# https://curl.haxx.se/download.html
+SRC_NAME=curl-7.54.1
-SRC_NAME=curl-7.37.x
 
@@ -12,9 +13,10 @@
 
 configure:
+	( test -d $(SRC_NAME) || tar -xzf $(SRC_NAME).tar.gz; \
+	cd $(SRC_NAME) ; \
-	( cd $(SRC_NAME) ; \
 	./configure \
 		--prefix=$(STAGEDIR) \
-		--enable-static \
-		--disable-shared \
+		--disable-static \
+		--enable-shared \
 		$(if $(CONFIG_IPV6),--enable-ipv6,--disable-ipv6) \
 		--disable-gopher \
@@ -59,4 +61,8 @@
 	$(MAKE) -C $(SRC_NAME) install DESTDIR=""
 
 romfs:
+ifeq ($(CONFIG_FIRMWARE_INCLUDE_CURL),y)
+	cp -fP $(DESTDIR)/bin/curl* $(ROMFSDIR)/usr/bin
+endif
+	cp -fP $(DESTDIR)/lib/libcurl.so* $(ROMFSDIR)/lib
 
